pkg/agent/controller/networkpolicy/node_reconciler_other.go

```diff
@@ -0,0 +1,769 @@
+	sync.Mutex
+}
+
+func newIPTRulesCache() *iptRulesCache {
+	return &iptRulesCache{
+		keyList: list.New(),
+		rules:   make(map[string]string),
+	}
+}
+
+// nodePolicyLastRealised is the struct cached by nodeReconciler. It's used to track the actual state of iptables rules and chains
+// we have enforced, so that we can know how to reconcile a rule when it's updated/removed.
+type nodePolicyLastRealised struct {
+	// ipsetNames tracks the last realized ipsets names for the policy rule.
+	ipsetNames map[iptables.Protocol]string
+	// serviceIPTChainName tracks the last realized service iptables chain.
+	serviceIPTChainName string
+	// coreIPTChainName tracks the last realized iptables chain where the core iptables rule is installed.
+	coreIPTChainName string
+	hasServices      bool
+	priority         *types.Priority
+}
+
+func newNodePolicyLastRealized() *nodePolicyLastRealised {
+	return &nodePolicyLastRealised{
+		ipsetNames: make(map[iptables.Protocol]string),
+	}
+}
+
+type nodeReconciler struct {
+	ipProtocols           []iptables.Protocol
+	routeClient           route.Interface
+	cachedCoreIPTRulesMap map[string]*iptRulesCache
+
+	// lastRealizeds caches the last realized rules. It's a mapping from ruleID to *nodePolicyLastRealised.
+	lastRealizeds sync.Map
+}
+
+func newNodeReconciler(routeClient route.Interface, ipv4Enabled, ipv6Enabled bool) (*nodeReconciler, error) {
+	var ipProtocols []iptables.Protocol
+	cachedCoreIPTRulesMap := make(map[string]*iptRulesCache)
+	if ipv4Enabled {
+		ipProtocols = append(ipProtocols, iptables.ProtocolIPv4)
+		cachedCoreIPTRulesMap[genCacheCategory(config.NodeNetworkPolicyIngressRulesChain, false)] = newIPTRulesCache()
+		cachedCoreIPTRulesMap[genCacheCategory(config.NodeNetworkPolicyEgressRulesChain, false)] = newIPTRulesCache()
+	}
+	if ipv6Enabled {
+		ipProtocols = append(ipProtocols, iptables.ProtocolIPv6)
+		cachedCoreIPTRulesMap[genCacheCategory(config.NodeNetworkPolicyIngressRulesChain, true)] = newIPTRulesCache()
+		cachedCoreIPTRulesMap[genCacheCategory(config.NodeNetworkPolicyEgressRulesChain, true)] = newIPTRulesCache()
+	}
+	return &nodeReconciler{
+		ipProtocols:           ipProtocols,
+		routeClient:           routeClient,
+		cachedCoreIPTRulesMap: cachedCoreIPTRulesMap,
+	}, nil
+}
+
+// Reconcile checks whether the provided rule have been enforced or not, and invoke the add or update method accordingly.
+func (r *nodeReconciler) Reconcile(rule *CompletedRule) error {
+	klog.InfoS("Reconciling Node NetworkPolicy rule", "rule", rule.ID, "policy", rule.SourceRef.ToString())
+
+	value, exists := r.lastRealizeds.Load(rule.ID)
+	var err error
+	if !exists {
+		err = r.add(rule)
+	} else {
+		err = r.update(value.(*nodePolicyLastRealised), rule)
+	}
+	return err
+}
+
+func (r *nodeReconciler) RunIDAllocatorWorker(stopCh <-chan struct{}) {
+
+}
+
+func (r *nodeReconciler) BatchReconcile(rules []*CompletedRule) error {
```

Comment:
I see duplicated code with `add`, can we extract some common code from `add` and lerverge the extracted func to construct the cache, then install them together?

Reply:
Will do it later.